# Vancouver Airbnb Listings - Consistency & Wrangling

This notebook contains:

1. Conducting basic checks on Airbnb Listings data set, including checking data types and descriptive statistics 
2. Data wrangling - dropping columns not required for analysis, renaming columns, etc.
2. Conducting data cleaning tasks - checking for missing values, duplicates, etc
3. Exporting cleaned data set as listings_clean

### Task 6.1 - Sourcing Open Data

### Importing Libraries and Dataframe

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [2]:
# Import dataset
df_listings = pd.read_csv(r'C:\Users\kaymi\OneDrive\Desktop\Career Foundry\Data Immersion\Achievement 6 - Advanced Data Analytics\02 Data\vancouver_listings.csv',index_col = False)

### Conducting Basic Checks

In [3]:
# Checking dimensions
df_listings.shape

(6355, 18)

In [4]:
# Checking columns and basic data
df_listings.head()

id                                               name  host_id  \
0  13188.0  Rental unit in Vancouver · ★4.83 · Studio · 2 ...    51466   
1  13358.0  Condo in Vancouver · ★4.68 · 1 bedroom · 1 bed...    52116   
2  13490.0  Rental unit in Vancouver · ★4.92 · 1 bedroom ·...    52467   
3  14267.0  Home in Vancouver · ★4.76 · 1 bedroom · 2 beds...    56030   
4  14424.0  Guest suite in Vancouver · ★4.69 · 1 bedroom ·...    56709   

        host_name  neighbourhood_group             neighbourhood  latitude  \
0     Family Guns                  NaN                Riley Park  49.24773   
1            Lynn                  NaN                  West End  49.28201   
2            Iris                  NaN  Kensington-Cedar Cottage  49.25622   
3  Peter & Alison                  NaN  Kensington-Cedar Cottage  49.24922   
4     Samantha Jo                  NaN         Downtown Eastside  49.27921   

   longitude        room_type  price  minimum_nights  number_of_reviews  \
0 -123.10509  Entire home/apt    150               3                260   
1 -123.12669  Entire home/apt    227               1                454   
2 -123.06607  Entire home/apt    150              30                 98   
3 -123.08139  Entire home/apt    150               3                 34   
4 -123.08835  Entire home/apt    134              30                266   

  last_review  reviews_per_month  calculated_host_listings_count  \
0  2023-06-01               1.61                               2   
1  2023-05-22               2.88                               1   
2  2023-05-01               0.67                               1   
3  2021-07-14               0.22                               1   
4  2023-05-25               1.64                               4   

   availability_365  number_of_reviews_ltm    license  
0               150                      4  22-156193  
1               247                     19  22-311727  
2               121                      5        NaN  
3                 0                      0  21-156500  
4               187                      7  19-162091

In [5]:
# Checking data types
df_listings.dtypes

id                                float64
name                               object
host_id                             int64
host_name                          object
neighbourhood_group               float64
neighbourhood                      object
latitude                          float64
longitude                         float64
room_type                          object
price                               int64
minimum_nights                      int64
number_of_reviews                   int64
last_review                        object
reviews_per_month                 float64
calculated_host_listings_count      int64
availability_365                    int64
number_of_reviews_ltm               int64
license                            object
dtype: object

In [8]:
# Checking descriptive statistics
df_listings.describe()

id       host_id  neighbourhood_group     latitude  \
count  6.355000e+03  6.355000e+03                  0.0  6355.000000   
mean   3.288925e+17  1.747044e+08                  NaN    49.263141   
std    3.807701e+17  1.695789e+08                  NaN     0.020564   
min    1.318800e+04  6.033000e+03                  NaN    49.202960   
25%    2.898755e+07  2.260310e+07                  NaN    49.251027   
50%    5.119770e+07  1.137402e+08                  NaN    49.269439   
75%    7.376940e+17  2.975667e+08                  NaN    49.279076   
max    9.095780e+17  5.186788e+08                  NaN    49.294360   

         longitude         price  minimum_nights  number_of_reviews  \
count  6355.000000   6355.000000     6355.000000        6355.000000   
mean   -123.112591    250.515972       16.147600          38.722738   
std       0.038044    571.811061       31.905929          69.252692   
min    -123.221859     14.000000        1.000000           0.000000   
25%    -123.131027    120.000000        2.000000           2.000000   
50%    -123.115937    179.000000        3.000000          11.000000   
75%    -123.090590    275.000000       30.000000          46.000000   
max    -123.023680  28386.000000      900.000000         888.000000   

       reviews_per_month  calculated_host_listings_count  availability_365  \
count        5248.000000                     6355.000000       6355.000000   
mean            1.773710                        6.200944        133.116601   
std             1.832335                       16.311660        118.588650   
min             0.010000                        1.000000          0.000000   
25%             0.320000                        1.000000         27.000000   
50%             1.030000                        1.000000         97.000000   
75%             2.830000                        3.000000        244.000000   
max            17.270000                      112.000000        365.000000   

       number_of_reviews_ltm  
count            6355.000000  
mean               13.159402  
std                20.025863  
min                 0.000000  
25%                 0.000000  
50%                 3.000000  
75%                19.000000  
max               147.000000

Listing price and minimum nights columns have a few very high values - requires further investigation. There also appears to be a host with 112 listings available which may be interesting to analyze.

### Data Wrangling

#### Dropping Columns

In [9]:
# Dropping neighbourhood_group and license columns
df_listings_wrangled = df_listings.drop(columns = ['neighbourhood_group','license'])

In [10]:
df_listings_wrangled.head()

id                                               name  host_id  \
0  13188.0  Rental unit in Vancouver · ★4.83 · Studio · 2 ...    51466   
1  13358.0  Condo in Vancouver · ★4.68 · 1 bedroom · 1 bed...    52116   
2  13490.0  Rental unit in Vancouver · ★4.92 · 1 bedroom ·...    52467   
3  14267.0  Home in Vancouver · ★4.76 · 1 bedroom · 2 beds...    56030   
4  14424.0  Guest suite in Vancouver · ★4.69 · 1 bedroom ·...    56709   

        host_name             neighbourhood  latitude  longitude  \
0     Family Guns                Riley Park  49.24773 -123.10509   
1            Lynn                  West End  49.28201 -123.12669   
2            Iris  Kensington-Cedar Cottage  49.25622 -123.06607   
3  Peter & Alison  Kensington-Cedar Cottage  49.24922 -123.08139   
4     Samantha Jo         Downtown Eastside  49.27921 -123.08835   

         room_type  price  minimum_nights  number_of_reviews last_review  \
0  Entire home/apt    150               3                260  2023-06-01   
1  Entire home/apt    227               1                454  2023-05-22   
2  Entire home/apt    150              30                 98  2023-05-01   
3  Entire home/apt    150               3                 34  2021-07-14   
4  Entire home/apt    134              30                266  2023-05-25   

   reviews_per_month  calculated_host_listings_count  availability_365  \
0               1.61                               2               150   
1               2.88                               1               247   
2               0.67                               1               121   
3               0.22                               1                 0   
4               1.64                               4               187   

   number_of_reviews_ltm  
0                      4  
1                     19  
2                      5  
3                      0  
4                      7

#### Renaming Columns

In [16]:
# Renaming id and name columns
df_listings_wrangled.rename(columns = {'id' : 'listing_id','name' : 'listing_name'}, inplace = True)

In [15]:
df_listings_wrangled.head()

listing_id                                       listing_name  host_id  \
0     13188.0  Rental unit in Vancouver · ★4.83 · Studio · 2 ...    51466   
1     13358.0  Condo in Vancouver · ★4.68 · 1 bedroom · 1 bed...    52116   
2     13490.0  Rental unit in Vancouver · ★4.92 · 1 bedroom ·...    52467   
3     14267.0  Home in Vancouver · ★4.76 · 1 bedroom · 2 beds...    56030   
4     14424.0  Guest suite in Vancouver · ★4.69 · 1 bedroom ·...    56709   

        host_name             neighbourhood  latitude  longitude  \
0     Family Guns                Riley Park  49.24773 -123.10509   
1            Lynn                  West End  49.28201 -123.12669   
2            Iris  Kensington-Cedar Cottage  49.25622 -123.06607   
3  Peter & Alison  Kensington-Cedar Cottage  49.24922 -123.08139   
4     Samantha Jo         Downtown Eastside  49.27921 -123.08835   

         room_type  price  minimum_nights  number_of_reviews last_review  \
0  Entire home/apt    150               3                260  2023-06-01   
1  Entire home/apt    227               1                454  2023-05-22   
2  Entire home/apt    150              30                 98  2023-05-01   
3  Entire home/apt    150               3                 34  2021-07-14   
4  Entire home/apt    134              30                266  2023-05-25   

   reviews_per_month  calculated_host_listings_count  availability_365  \
0               1.61                               2               150   
1               2.88                               1               247   
2               0.67                               1               121   
3               0.22                               1                 0   
4               1.64                               4               187   

   number_of_reviews_ltm  
0                      4  
1                     19  
2                      5  
3                      0  
4                      7

#### Changing Data Types

In [20]:
# Changing listing_id from float to int
df_listings_wrangled['listing_id'] = df_listings_wrangled['listing_id'].astype('int64')

In [23]:
df_listings_wrangled['listing_id'].dtype

dtype('int64')

In [24]:
df_listings_wrangled.head()

listing_id                                       listing_name  host_id  \
0       13188  Rental unit in Vancouver · ★4.83 · Studio · 2 ...    51466   
1       13358  Condo in Vancouver · ★4.68 · 1 bedroom · 1 bed...    52116   
2       13490  Rental unit in Vancouver · ★4.92 · 1 bedroom ·...    52467   
3       14267  Home in Vancouver · ★4.76 · 1 bedroom · 2 beds...    56030   
4       14424  Guest suite in Vancouver · ★4.69 · 1 bedroom ·...    56709   

        host_name             neighbourhood  latitude  longitude  \
0     Family Guns                Riley Park  49.24773 -123.10509   
1            Lynn                  West End  49.28201 -123.12669   
2            Iris  Kensington-Cedar Cottage  49.25622 -123.06607   
3  Peter & Alison  Kensington-Cedar Cottage  49.24922 -123.08139   
4     Samantha Jo         Downtown Eastside  49.27921 -123.08835   

         room_type  price  minimum_nights  number_of_reviews last_review  \
0  Entire home/apt    150               3                260  2023-06-01   
1  Entire home/apt    227               1                454  2023-05-22   
2  Entire home/apt    150              30                 98  2023-05-01   
3  Entire home/apt    150               3                 34  2021-07-14   
4  Entire home/apt    134              30                266  2023-05-25   

   reviews_per_month  calculated_host_listings_count  availability_365  \
0               1.61                               2               150   
1               2.88                               1               247   
2               0.67                               1               121   
3               0.22                               1                 0   
4               1.64                               4               187   

   number_of_reviews_ltm  
0                      4  
1                     19  
2                      5  
3                      0  
4                      7

### Data Consistency Checks

#### Checking for Mixed Type Data

In [17]:
# Checking for mixed-type data
for col in df_listings_wrangled.columns.tolist():
  weird = (df_listings_wrangled[[col]].applymap(type) != df_listings_wrangled[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_listings_wrangled[weird]) > 0:
    print (col)

last_review


In [18]:
# Update last_review column data type to string
df_listings_wrangled['last_review'] = df_listings_wrangled['last_review'].astype('str')

In [19]:
for col in df_listings_wrangled.columns.tolist():
  weird = (df_listings_wrangled[[col]].applymap(type) != df_listings_wrangled[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_listings_wrangled[weird]) > 0:
    print (col)

last_review mixed column type corrected

#### Checking for Missing Values

In [25]:
df_listings_wrangled.isnull().sum()

listing_id                           0
listing_name                         0
host_id                              0
host_name                            0
neighbourhood                        0
latitude                             0
longitude                            0
room_type                            0
price                                0
minimum_nights                       0
number_of_reviews                    0
last_review                          0
reviews_per_month                 1107
calculated_host_listings_count       0
availability_365                     0
number_of_reviews_ltm                0
dtype: int64

In [26]:
# View missing values in reviews_per_month column
df_nan = df_listings_wrangled[df_listings_wrangled['reviews_per_month'].isnull() == True]

In [27]:
df_nan

listing_id                                       listing_name  \
9                  18024  Home in Vancouver · 3 bedrooms · 3 beds · 1 pr...   
71                754118  Rental unit in Vancouver · 2 bedrooms · 2 beds...   
79                796603  Rental unit in Vancouver · 1 bedroom · 1 bed ·...   
165              2374164  Home in Vancouver · ★New · 3 bedrooms · 3 beds...   
190              2876573  Rental unit in Vancouver · 2 bedrooms · 1 bed ...   
...                  ...                                                ...   
6350  909442000000000000  Rental unit in Vancouver · ★New · 2 bedrooms ·...   
6351  909450000000000000  Home in Vancouver · ★New · 1 bedroom · 1 bed ·...   
6352  909475000000000000  Rental unit in Vancouver · ★New · 1 bedroom · ...   
6353  909539000000000000  Home in Vancouver · ★New · 1 bedroom · 1 bed ·...   
6354  909578000000000000  Rental unit in Vancouver · ★New · 3 bedrooms ·...   

        host_id host_name             neighbourhood   latitude   longitude  \
9         69472     Raine  Kensington-Cedar Cottage  49.247810 -123.072410   
71      3967530      Eric  Kensington-Cedar Cottage  49.249376 -123.089920   
79      4196444   Milorad                  West End  49.280870 -123.135250   
165    12067546     Mitch  Kensington-Cedar Cottage  49.261926 -123.066234   
190       52926   Shohreh                  Downtown  49.289450 -123.126650   
...         ...       ...                       ...        ...         ...   
6350  474073067       Zoe                  Downtown  49.275740 -123.115920   
6351   82112046     Allie               Shaughnessy  49.239280 -123.149150   
6352  227662329    Jordan            Mount Pleasant  49.266040 -123.097260   
6353   31047232      Gwen                Kerrisdale  49.230884 -123.163359   
6354  287607431    Elliot                  Downtown  49.274960 -123.127950   

            room_type  price  minimum_nights  number_of_reviews last_review  \
9        Private room    150              30                  0         nan   
71    Entire home/apt    116              30                  0         nan   
79        Shared room     30              30                  0         nan   
165   Entire home/apt    249               4                  0         nan   
190   Entire home/apt    160              30                  0         nan   
...               ...    ...             ...                ...         ...   
6350  Entire home/apt    500               5                  0         nan   
6351     Private room     44              30                  0         nan   
6352  Entire home/apt     82              30                  0         nan   
6353     Private room     45               1                  0         nan   
6354  Entire home/apt    599              30                  0         nan   

      reviews_per_month  calculated_host_listings_count  availability_365  \
9                   NaN                               1                 0   
71                  NaN                               1                36   
79                  NaN                               1                 0   
165                 NaN                               1                23   
190                 NaN                               1               184   
...                 ...                             ...               ...   
6350                NaN                              16               364   
6351                NaN                               1               179   
6352                NaN                             112               363   
6353                NaN                               2               106   
6354                NaN                               3               362   

      number_of_reviews_ltm  
9                         0  
71                        0  
79                        0  
165                       0  
190                       0  
...                     ...  
6350                      0  
6351          

It appears that the missing values in the reviews per month column are associated with any listing that has zero reviews - I have decided to leave this as is for now as an indicator to myself that these listings have no review information.

#### Checking for Duplicates

In [28]:
#Checking for duplicates
df_dups = df_listings_wrangled[df_listings_wrangled.duplicated()]

In [29]:
df_dups

Empty DataFrame
Columns: [listing_id, listing_name, host_id, host_name, neighbourhood, latitude, longitude, room_type, price, minimum_nights, number_of_reviews, last_review, reviews_per_month, calculated_host_listings_count, availability_365, number_of_reviews_ltm]
Index: []

No duplicates found

### Exporting Dataframe

In [30]:
path = r'C:\Users\kaymi\OneDrive\Desktop\Career Foundry\Data Immersion\Achievement 6 - Advanced Data Analytics'

In [31]:
df_listings_wrangled.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'listings_clean.csv'))